src/org/etools/j1939_84/controllers/part01/Part01Step26Controller.java

```diff
@@ -300,47 +307,618 @@ protected void run() throws Throwable {
+                                                                             // Record each
+                                                                             // value
+                                                                             // for use in
+                                                                             // Part 2.
+                                                                             .peek(this::save)
+                                                                             .collect(Collectors.toList());
+
+        if (!nOx100HourPackets.isEmpty()) {
+            // 6.1.26.9.c - List data received in a table using bin numbers for rows.
+            getListener().onResult(nOxBinningModule.format(nOx100HourPackets));
+        }
+        for (int pg : nOx100HourSps) {
+            GenericPacket packetForPg = haveResponseWithPg(nOx100HourPackets, pg);
+            if (packetForPg == null) {
+                // 6.1.26.10.a. For all MY2024+ Diesel engines, Fail each PG query where no response was received.
+                if (getEngineModelYear() >= 2024) {
+                    addFailure("6.1.26.10.a - No response was received from "
+                            + module.getModuleName() + "for PG "
+                            + pg);
+                }
+                // 6.1.26.10.b. For all MY2022-23 Diesel engines, Warn each PG query where no response was received.
+                if (2022 <= getEngineModelYear() && getEngineModelYear() >= 2023) {
+                    addWarning("6.1.26.10.b - No response was received from "
+                            + module.getModuleName() + "for PG "
+                            + pg);
+                }
+            } else {
+                packetForPg.getSpns().forEach(spn -> {
+                    if (spn.getRawValue() >= 0xFAFFFFFFL) {
+                        // 6.1.26.10.c. Fail each PG query where any bin value received is greater than FAFFh. (Use
+                        // FAFFFFFFh for NOx values)
+                        addFailure("6.1.26.10.c - Bin value received is greater than 0xFAFFFFFF(h)"
+                                + module.getModuleName() + " returned " + Arrays.toString(spn.getBytes()) + " for "
+                                + spn);
+
+                    }
+                    // 6.1.26.10.d. Fail each active 100 hr array value that is greater than zero. (where supported)
+                    if (spn.getValue() > 0) {
+                        // 6.1.26.10.d. Fail each active 100 hr array value that is greater than zero. (where supported)
+                        addFailure("6.1.26.10.d - Active 100 hr array value received is greater than zero (where supported)"
+                                + module.getModuleName() + " returned " + Arrays.toString(spn.getBytes()) + " for "
+                                + spn);
+
+                    }
+                });
+            }
+        }
```

Comment:
All this GHG checking should probably be in a separate class, just because there is SOOOOO much code.  It’d be great if we could have one or two functions to do all this checking as it seems to be rather repeative.